### DISCUSSION TITLE
Replay Structure

### DISCUSSION THREAD

**Discussion by CataclysmicCode** [eigenein/wotbreplay-parser#3] [Q&A] [answered]
I have little to no understanding of rust so I’m having trouble reading the code solve my problem. All I’m wondering is how you figured out the structure of the replay files in the first place and if it might be possible to modify and then repack the file when done. (So you can update old replay files when the world of tanks blitz changes their client)

**Comment by CataclysmicCode** [comment]
I’m trying to do it in php and I’ve made a little progress, but I still need to know the structure. Would any one of these fit?


a - NUL-padded string
A - SPACE-padded string
h - Hex string, low nibble first
H - Hex string, high nibble first
c - signed char
C - unsigned char
s - signed short (always 16 bit, machine byte order)
S - unsigned short (always 16 bit, machine byte order)
n - unsigned short (always 16 bit, big endian byte order)
v - unsigned short (always 16 bit, little endian byte order)
i - signed integer (machine dependent size and byte order)
I - unsigned integer (machine dependent size and byte order)
l - signed long (always 32 bit, machine byte order)
L - unsigned long (always 32 bit, machine byte order)
N - unsigned long (always 32 bit, big endian byte order)
V - unsigned long (always 32 bit, little endian byte order)
q - signed long long (always 64 bit, machine byte order)
Q - unsigned long long (always 64 bit, machine byte order)
J - unsigned long long (always 64 bit, big endian byte order)
P - unsigned long long (always 64 bit, little endian byte order)
f - float (machine dependent size and representation)
g - float (machine dependent size, little endian byte order)
G - float (machine dependent size, big endian byte order)
d - double (machine dependent size and representation)
e - double (machine dependent size, little endian byte order)
E - double (machine dependent size, big endian byte order)
x - NUL byte
X - Back up one byte
Z - NUL-padded string
@ - NUL-fill to absolute

**Reply by eigenein** [reply]
Besides the «normal» fixed-length formats, you'll need to deserialize Pickle and Protocol Buffers

**Comment by eigenein** [comment]
> All I’m wondering is how you figured out the structure of the replay files in the first place

Some fields I've reversed by hand, some were suggested by other developers from other Blitz-related projects, and some I've found in other repo's on GitHub and BitBucket (with a lot of googling). It's really a combination of different pieces here and there

> if it might be possible to modify and then repack the file when done

Well, it's surely possible to change and repack it back by writing the data back in the same format. This parser, however, loses tags that it «doesn't know». You'd probably need to use some protobuf parser that doesn't lose unknown tags

> So you can update old replay files when the world of tanks blitz changes their client

That depends on how much they'd change it. E.g. if they were to add some new fields that are missing in the current version, you'd need to figure out how to fill them in. If they change the serialization format, you'd need to adapt. Etc, it might get challenging

**Comment by CataclysmicCode** [comment]
So, it sounds as though the structure is quite complex then and I’m guessing that it probably can’t be unpacked in a just few lines or anything.

Do you mind explaining what a protobuf parser is? I’m just getting the hang of data handling with my expertise being more placed in other areas.

Yes, that was the general idea, but as you say that may be above my pay grade. It would have been nice if wargaming just put a little documentation out there XD

**Reply by eigenein** [reply, marked as answer]
> Do you mind explaining what a protobuf parser is?

If it were that easy :D Here's the complete docs: https://protobuf.dev

> It would have been nice if wargaming just put a little documentation out there

💯 For everyone, _including Wargaming_
